### ISSUE TITLE
next-generation p2p protocol

### ISSUE THREAD

**Issue by cc32d9** [eosnetworkfoundation/mandel#125]
This is a wishlist for a next generation p2p protocol that may eventually be implemented.

**P2PW001**.  In current p2p protocol, if a node receives a message of unknown type, it can't even ignore it, because it doesn't know the message length. The message encoding should include its length, so that the node could ignore the messages that it cannot interpret.

**Comment by matthewdarwin** [comment]
- supports ipv6
- works over web socket or something that doesn't require dedicated IP address/port firewall rules
- improved handling on forks

**Comment by spoonincode** [comment]
> supports ipv6

fwiw there isn't anything in the current protocol that prohibits this from being added.

**Comment by heifner** [comment]
> > supports ipv6
> 
> fwiw there isn't anything in the current protocol that prohibits this from being added.

Just the few hard-coded ipv4 calls like: https://github.com/eosnetworkfoundation/mandel/blob/main/plugins/net_plugin/net_plugin.cpp#L2254

**Comment by cc32d9** [comment]
Todd mentioned some difficulties in Boost library, but I don't know the details. But IPv6 would definitely be a nice thing.

**Comment by spoonincode** [comment]
* by default nodeos should maintain outgoing connections to a limited number of peers. Users tend to copy paste dozens of p2p servers in to their config causing excessive burden on the network as a whole.
* peer discovery

**Comment by matthewdarwin** [comment]
https://github.com/EOSIO/eos/issues/8170

**Comment by matthewdarwin** [comment]
https://github.com/EOSIO/eos/issues/8171

**Comment by matthewdarwin** [comment]
https://github.com/EOSIO/eos/issues/8260

**Comment by matthewdarwin** [comment]
https://github.com/EOSIO/eos/issues/8442

**Comment by matthewdarwin** [comment]
https://github.com/EOSIO/eos/issues/8821

**Comment by matthewdarwin** [comment]
https://docs.libp2p.io/introduction/what-is-libp2p/

**Comment by matthewdarwin** [comment]
> Todd mentioned some difficulties in Boost library, but I don't know the details. But IPv6 would definitely be a nice thing.

https://github.com/EOSIO/eos/issues/8169

**Comment by spoonincode** [comment]
Yes IME IPv6 troubles with asio is (as often the case with asio) asio being pedantic and pretty raw.

For example the existing http plugin when setting up the listening socket does something like
```c++
tcp::resolver::query query( tcp::v4(), host, port);
my->listen_endpoint = *resolver.resolve( query );
```
Sure, that works for IPv4 today. But what if I wanted to support IPv6 too? Well I could do something like
```c++
tcp::resolver::query query(host /* "localhost" */, port);
my->listen_endpoint = *resolver.resolve( query );
```
**BUG**: the resolve here will actually return two endpoints: one with `127.0.0.1` and one with `::1`. So application code needs to be smart enough to iterate through and listen on all of them.

For asio that abstracts away the address family nicely, that's actually usually enough. Without asio, if you're writing closer to BSD sockets, IMO it's nice to always use IPv6 family across the board. Just for illustration purposes, you might emulate something like that in asio with
```c++
resolver.resolve(tcp::v6(), "localhost", "5555", tcp::resolver::v4_mapped | tcp::resolver::all_matching);
```
This will return` ::1` and `::ffff:127.0.0.1` But wait! **BUG**: `IPV6_V6ONLY` is set off by default on Linux and on by default on MacOS. So trying to listen on `::ffff:127.0.0.1` will error out on MacOS, I believe, unless you go and set `boost::asio::ip::v6_only` too!

We've only talked about listening sockets; outgoing sockets is a whole other ball game. You'll want to read up on "Happy Eyeballs" -- [RFC6555](https://www.rfc-editor.org/rfc/rfc6555) & [RFC8305](https://www.rfc-editor.org/rfc/rfc8305). cleos, imo needs to follow these suggestions. It's another reason on top of #110 I'm suggesting moving to the platform HTTP provider for cleos. Something like libcurl, WinHTTP, NSURLSession, all already have lots of smarts like Happy Eyeballs built in.

p2p _probably_ doesn't need Happy Eyeballs because it's probably acceptable to round robin resolved addresses after they naturally time out.

Anyways, mostly want to point out that properly supporting IPv6 does require a little love beyond just flipping something in the code.

**Comment by jgiszczak** [comment]
Swarm download.

**Comment by matthewdarwin** [comment]
BPs run multiple public p2p endpoint as a cluster of nodeos together behind a load balancer. Today all nodeos have to have all the blocks. Ideally all the "older" blocks are not stored on all nodes, but can retrieve them from another node to send them out.  Eg for block producers they might run 10 p2p servers. Most clients are asking about the last few blocks only so having every node in the cluster have all blocks is waste of disk space.  Could be done via some sort of smart proxy?

**Comment by matthewdarwin** [comment]
Ability to start from snapshot via p2p.

**Comment by cc32d9** [comment]
> Ability to start from snapshot via p2p.

this will require your neighbor to freeze operation until you download the state. Not something applicable to public peers

**Comment by roger-eosuk** [comment]
Sure will drop the link to Impervious the p2p method over Bitcoin Lightning network on Bitcoin. 
https://www.impervious.ai/ 
https://twitter.com/ImperviousAi 
https://docs.impervious.ai/ 
https://www.publicdomain.live/Impervious 
This is a great demo video https://www.youtube.com/watch?v=O0Dcx6ScnRY 
So in theory what they do for 15sats in the demo we could do for free on eosio, right?

**Comment by jgiszczak** [comment]
@roger-eosuk Impervious is orthogonal to this thread.  This thread is about the protocol nodes would use which are participating in the network that actually produces and propagates the chain.

Having said that, yes, a version of Impervious could be implemented on eosio, and it could operate for free, requiring only sufficient staking (which is refundable) for the activity of each user.  It could be done now, implemented entirely in smart contracts.  No changes are required to the code the nodes run.

**Comment by sunburntcat** [comment]
Ability to begin sync from specific irreversible block

**Comment by tedcahalleos** [comment]
Per KevinH, it is now multi-threaded and stable performance is great.  But the syncing performance and algorithms need to be rewritten.  LIB catch-up, Head catch up, Synchonize, transition of those barriers is not healthy.  Not micro-fork aware.  Some awareness helps it not redo work (re-requesting and re-fetching blocks).  Likely needs queues to hold data it will need again.  The current state machine is convoluted.

**Comment by matthewdarwin** [comment]
Don't connect to p2p peers while replaying blocks.log

**Comment by matthewdarwin** [comment]
Remove the feature to have public/private key for connecting p2p peers.

**Comment by matthewdarwin** [comment]
> > Ability to start from snapshot via p2p.
> 
> this will require your neighbor to freeze operation until you download the state. Not something applicable to public peers

If your neighbour generates snapshots on a regular basis, then you just get the last one.

**Comment by matthewdarwin** [comment]
The new "next-gen" p2p protocol should run in parallel to (on a different port) to the existing p2p protocol.

**Comment by matthewdarwin** [comment]
The goal should be to allow BPs to have same control they have today, but for the average developer, they should be able to just start nodeos with a static configuration and things just work without messing with peers and snapshots and whatever else.

**Comment by DenisCarriere** [comment]
Perhaps [gRPC](https://grpc.io/) instead of Websocket ?

@matthewdarwin 
> * supports ipv6
> * works over web socket or something that doesn't require dedicated IP address/port firewall rules
> * improved handling on forks

**Comment by sunburntcat** [comment]
Title:

Move transaction headers (TaPOS) into the p2p protocol

Background:
Block head number and block prefix are currently required for every eosio transaction. This was done to improve fork handling and prevent double-spend of a transaction as it moves around the p2p network. However, this means end users need to query a network RPC node to get this information before they submit a transaction... a very time intensive process.

Suggested solution:
Remove this requirement and instead allow RPC nodes to append these TaPOS fields to a broadcasted transaction. And then append their own signature before they propagate it onto the next peer. Ideally, it could be implemented as a --signature-provider argument inside http plugin in nodeos.

Expected Outcomes:
- Significant reduction in overhead and latency for client libraries like eosjs.
- Opens the door to sending transactions from devices with poor network connection, including sensors wanting to broadcast transactions to the chain.

**Comment by swatanabe** [comment]
> Move transaction headers (TaPOS) into the p2p protocol

This won't work.  tapos in the p2p protocol cannot fulfill the same role as tapos in transaction signatures.  Also, getting tapos doesn't need to be expensive.  It's only slow because the API did not provide a lightweight interface for getting it.

**Comment by sunburntcat** [comment]
The point is still there that ideally a transaction should not need ref_block_num and ref_block_prefix if the RPC they are submitting to already has this information.

If a transaction, for whatever reason, was meant to be applicable to a particular fork of the chain (implied by num/prefix), then they could simply submit their transaction to a different node on that fork.
